### ISSUE TITLE
runner can hang after OOM

### ISSUE THREAD

**Issue by mlin** [chanzuckerberg/miniwdl#263]
Saw one time -- opening for tracking

* kernel OOM killer did its thing
* miniwdl verbose log displayed "docker task state = shutdown" and appeared to get stuck there
* docker service ps showed 0/1 replicates, docker service inspect xxx did not show any container or exit code
* miniwdl responded to ctrl-C and exited

**Comment by mlin** [comment]
Pushed possible fix -- hypothesized cause was unrelated to OOM, except perhaps that being in a low-memory state changed performance characteristics to lose a race condition

**Comment by mlin** [comment]
Problem still exists. It seems that, even short of OOM killer invocation, if the system performance is erratic due to memory swapping, then dockerd can get confused and leave a swarm task wedged in the "Shutdown" state. Seen in `/var/log/syslog`:

```
Nov 10 18:45:31 BUFFX rtkit-daemon[1334]: The canary thread is apparently starving. Taking action.
Nov 10 18:45:34 BUFFX dockerd[17881]: time="2019-11-10T18:45:31.358895717-10:00" level=info msg="NetworkDB stats BUFFX(d116a2effb40) - netID:setc4orfslnksy6f5t3wa91zt leaving:false netPeers:1 entries:2 Queue qLen:0 netMsg/s:0"
Nov 10 18:45:34 BUFFX dockerd[17881]: time="2019-11-10T18:45:32.835022636-10:00" level=error msg="heartbeat to manager { } failed" error="rpc error: code = NotFound desc = node not registered" method="(*session).heartbeat" module=node/agent node.id=nikibxav5jj6n4nfm7vl25uwu session.id=gx5wnqfjrat84ewrzn99x2iz3 sessionID=gx5wnqfjrat84ewrzn99x2iz3
Nov 10 18:45:34 BUFFX dockerd[17881]: time="2019-11-10T18:45:33.362833032-10:00" level=error msg="agent: session failed" backoff=100ms error="node not registered" module=node/agent node.id=nikibxav5jj6n4nfm7vl25uwu
Nov 10 18:45:34 BUFFX dockerd[17881]: time="2019-11-10T18:45:33.380745692-10:00" level=info msg="manager selected by agent for new session: { }" module=node/agent node.id=nikibxav5jj6n4nfm7vl25uwu
Nov 10 18:45:34 BUFFX dockerd[17881]: time="2019-11-10T18:45:33.387825533-10:00" level=info msg="waiting 42.494484ms before registering session" module=node/agent node.id=nikibxav5jj6n4nfm7vl25uwu
Nov 10 18:45:34 BUFFX dockerd[17881]: sync duration of 1.628562736s, expected less than 1s
Nov 10 18:45:34 BUFFX rtkit-daemon[1334]: Demoting known real-time threads.
Nov 10 18:45:34 BUFFX dockerd[17881]: time="2019-11-10T18:45:34.239917032-10:00" level=info msg="worker nikibxav5jj6n4nfm7vl25uwu was successfully registered" method="(*Dispatcher).register"
Nov 10 18:45:34 BUFFX rtkit-daemon[1334]: Successfully demoted thread 3022 of process 2816.
Nov 10 18:45:34 BUFFX rtkit-daemon[1334]: Successfully demoted thread 3021 of process 2816.
Nov 10 18:45:34 BUFFX rtkit-daemon[1334]: Successfully demoted thread 2817 of process 2816.
Nov 10 18:45:34 BUFFX rtkit-daemon[1334]: Successfully demoted thread 2816 of process 2816.
Nov 10 18:45:34 BUFFX rtkit-daemon[1334]: Demoted 4 threads.
Nov 10 18:45:45 BUFFX dockerd[17881]: time="2019-11-10T18:45:45.795576949-10:00" level=info msg="Container e7d1c8b7a662951d8d75761c5d6b35807971d83a066acabcc41c3ade51f1667c failed to exit within 10 seconds of signal 15 - using the force"
Nov 10 18:45:47 BUFFX containerd[1149]: time="2019-11-10T18:45:47.483474693-10:00" level=info msg="shim reaped" id=e7d1c8b7a662951d8d75761c5d6b35807971d83a066acabcc41c3ade51f1667c
Nov 10 18:45:47 BUFFX dockerd[17881]: time="2019-11-10T18:45:47.499898409-10:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
```
We still see this task hanging around:

```
$ docker node ps
ID                  NAME                IMAGE                                                             NODE                DESIRED STATE       CURRENT STATE          ERROR               PORTS
2416gbnsgw0c        lucid_satoshi.1     quay.io/humancellatlas/secondary-analysis-hisat2:v0.2.2-2-2.1.0   BUFFX               Shutdown            Shutdown 2 hours ago                      
```
We *do* have ContainerStatus.ExitCode:

```
$ docker inspect 2416gbnsgw0c | jq .[0].Status
{
  "Timestamp": "2019-11-11T04:45:47.845632944Z",
  "State": "shutdown",
  "Message": "shutdown",
  "ContainerStatus": {
    "ContainerID": "e7d1c8b7a662951d8d75761c5d6b35807971d83a066acabcc41c3ade51f1667c",
    "PID": 0,
    "ExitCode": 137
  },
  "PortStatus": {}
}
```

so maybe we can workaround by waiting for that instead of the complete/failed state.

**Comment by mlin** [comment]
second possible fix - 4bf26353866999d7137cf1d495282ba93dbf8aaf

**Comment by mlin** [comment]
Reasonably confident in that fix, so closing unless/until we see this again
